packages/eip-7702-internal-rpc-middleware/src/hooks/upgradeAccount.ts

```diff
@@ -0,0 +1,113 @@
+import { toHex } from '@metamask/controller-utils';
+import { rpcErrors } from '@metamask/rpc-errors';
+import { tuple } from '@metamask/superstruct';
+import type {
+  JsonRpcRequest,
+  PendingJsonRpcResponse,
+  Json,
+} from '@metamask/utils';
+
+import type {
+  UpgradeAccountParams,
+  UpgradeAccountResult,
+  UpgradeAccountHooks,
+} from '../types';
+import { UpgradeAccountParamsStruct } from '../types';
+import { validateParams, validateAndNormalizeAddress } from '../utils';
+
+/**
+ * Upgrades an EOA account to a smart account using EIP-7702.
+ *
+ * @param req - The JSON-RPC request object containing the upgrade parameters.
+ * @param _res - The JSON-RPC response object (unused).
+ * @param options - Configuration object containing required functions.
+ * @param options.upgradeAccount - Function to perform the account upgrade.
+ * @param options.getCurrentChainIdForDomain - Function to get the current chain ID for a domain.
+ * @param options.isEip7702Supported - Function to check if EIP-7702 is supported.
+ * @param options.getAccounts - Function to get accounts for the requester.
+ * @returns Promise that resolves to the upgrade result containing transaction hash and delegation info.
+ */
+export async function upgradeAccount(
+  req: JsonRpcRequest<Json[]> & { origin: string },
+  _res: PendingJsonRpcResponse,
+  {
+    upgradeAccount: upgradeAccountFn,
+    getCurrentChainIdForDomain,
+    isEip7702Supported,
+    getAccounts,
+  }: UpgradeAccountHooks,
+): Promise<UpgradeAccountResult> {
+  const { params, origin } = req;
+
+  // Validate parameters using Superstruct
+  validateParams(params, tuple([UpgradeAccountParamsStruct]));
+
+  const [upgradeParams] = params as [UpgradeAccountParams];
+  const { account, chainId } = upgradeParams;
+
+  // Validate and normalize the account address with authorization check
+  const normalizedAccount = await validateAndNormalizeAddress(account, req, {
+    getAccounts,
+  });
+
+  // Use current chain ID if not provided
+  let targetChainId: number;
+  if (chainId !== undefined) {
+    targetChainId = chainId;
+  } else {
+    const currentChainIdForDomain = getCurrentChainIdForDomain(origin);
+    if (!currentChainIdForDomain) {
+      throw rpcErrors.invalidParams({
+        message: `No network configuration found for origin: ${origin}`,
+      });
+    }
+    targetChainId = parseInt(currentChainIdForDomain, 16);
```

Comment:
`@metamask/controller-utils` has a `fromHex` function, should we use that instead?